core/src/schema.rs

```diff
@@ -0,0 +1,112 @@
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use ::{
+    Attribute,
+    Entid,
+    Schema,
+};
+use edn::NamespacedKeyword;
+
+impl Schema {
+    pub fn get_ident(&self, x: Entid) -> Option<&NamespacedKeyword> {
+        self.entid_map.get(&x)
+    }
+
+    pub fn get_entid(&self, x: &NamespacedKeyword) -> Option<Entid> {
+        self.ident_map.get(x).map(|x| *x)
+    }
+
+    pub fn attribute_for_entid(&self, x: Entid) -> Option<&Attribute> {
+        self.schema_map.get(&x)
+    }
+
+    pub fn attribute_for_ident(&self, ident: &NamespacedKeyword) -> Option<&Attribute> {
+        self.get_entid(&ident)
+            .and_then(|x| self.attribute_for_entid(x))
+    }
+
+    /// Return `true` if the provided entid identifies an attribute in this schema.
+    pub fn is_attribute(&self, x: Entid) -> bool {
+        self.schema_map.contains_key(&x)
+    }
+
+    /// Return `true` if the provided ident identifies an attribute in this schema.
+    pub fn identifies_attribute(&self, x: &NamespacedKeyword) -> bool {
+        self.get_entid(x).map(|e| self.is_attribute(e)).unwrap_or(false)
+    }
+
+    /// Associate the given ident with given entid if neither are already recognized.
+    ///
+    /// Return `true` if an association was added, `false` otherwise.
+    pub fn associate_ident(&mut self, i: NamespacedKeyword, e: Entid) -> bool {
+        use std::collections::btree_map::Entry;
+        match (self.entid_map.entry(e), self.ident_map.entry(i.clone())) {
+            (Entry::Vacant(entid_entry), Entry::Vacant(ident_entry)) => {
+                entid_entry.insert(i);
+                ident_entry.insert(e);
+                true
+            },
+            _ => false,
+        }
+    }
+
+    /// Dissociate the given ident and the given entid if the two are associated.
+    ///
+    /// Return `true` if an association was removed, `false` otherwise.
+    pub fn dissociate_ident(&mut self, i: NamespacedKeyword, e: Entid) -> bool {
+        use std::collections::btree_map::Entry;
+        match (self.entid_map.entry(e), self.ident_map.entry(i.clone())) {
+            (Entry::Occupied(entid_entry), Entry::Occupied(ident_entry)) => {
+                if entid_entry.get() == &i && ident_entry.get() == &e {
+                    entid_entry.remove();
+                    ident_entry.remove();
+                    true
+                } else {
+                    false
+                }
+            },
+            _ => false,
+        }
+    }
+
+    /// XXX
+    pub fn update_ident(&mut self, i: NamespacedKeyword, e: Entid) -> bool {
+        use std::collections::btree_map::Entry;
+        if let Entry::Occupied(mut entid_entry) = self.entid_map.entry(e) {
+            // TODO: avoid clone?
```

Comment:
To avoid the clone here you can use the return value of `remove` to decide whether to return false. You don't do anything with the second entry beyond making sure that it's occupied before removing it; `remove` returns `None` if there was nothing to remove.